File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        c1, c2, c = l1, l2, 0
        dummy = ListNode(-1)
        tail = dummy
        while c1 or c2 or c:
            v1 = c1.val if c1 else 0
            v2 = c2.val if c2 else 0
            tot = v1 + v2 + c
            node = ListNode(tot%10)
            c = tot//10
            tail.next = node
            tail = node
            if c1:
                c1 = c1.next
            if c2:
                c2 = c2.next
                
        return dummy.next
            
        

File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode p = l1, q = l2;
        int c = 0;

        ListNode d = new ListNode(-1), t = d;

        while (p != null || q!=null || c > 0 ) {
            int pval = p != null ? p.val : 0;
            int qval = q != null ? q.val : 0;

            int sum = pval + qval + c;
            ListNode n = new ListNode(sum % 10);
            t.next = n;
            t = n;
            c =  (int) sum / 10;
            if (p != null) p = p.next;
            if (q != null) q = q.next;
        } 

        if (c > 0) {
            t.next = new ListNode(1);
        }

        return d.next;
    }
}